structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2017 Google Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drivers
+
+import (
+	"bufio"
+	"bytes"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/runtimes-common/structure_tests/types/unversioned"
+	"github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	cli           docker.Client
+	originalImage string
+	currentImage  string
+}
+
+func NewDockerDriver(image string) Driver {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		panic(err)
+	}
+	return &DockerDriver{
+		originalImage: image,
+		currentImage:  image,
+		cli:           *newCli,
+	}
+}
+
+func (d *DockerDriver) Setup(t *testing.T, envVars []unversioned.EnvVar, fullCommand []unversioned.Command) {
+	env := d.processEnvVars(t, envVars)
+	for _, cmd := range fullCommand {
+		d.currentImage = d.runAndCommit(t, env, cmd)
+	}
+}
+
+func (d *DockerDriver) ProcessCommand(t *testing.T, envVars []unversioned.EnvVar, fullCommand []string,
+	checkOutput bool) (string, string, int) {
+
+	var env []string
+	for _, envVar := range envVars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	//reset image for next test
+	d.currentImage = d.originalImage
+	return stdout, stderr, exitCode
+}
+
+func (d *DockerDriver) processEnvVars(t *testing.T, vars []unversioned.EnvVar) []string {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	image, err := d.cli.InspectImage(d.currentImage)
+	if err != nil {
+		t.Errorf("Error when inspecting image: %s", err.Error())
+		return nil
+	}
+
+	// convert env to map for easier processing
+	imageEnv := make(map[string]string)
+	for _, varPair := range image.Config.Env {
+		pair := strings.Split(varPair, "=")
+		imageEnv[pair[0]] = pair[1]
+	}
+
+	before := regexp.MustCompile(".*\\$(.*?):")
```

Comment:
could you leave a comment explaining what this regex is supposed to match?

Reply:
done